Shutdown race

```2020-01-22 16:05:08.285 [INF] DCRD: Received signal (interrupt).  Shutting down...
2020-01-22 16:05:08.285 [WRN] SRVR: Server shutting down
2020-01-22 16:05:08.285 [WRN] RPCS: RPC server shutting down
2020-01-22 16:05:08.291 [INF] RPCS: RPC server shutdown complete
2020-01-22 16:05:08.292 [INF] BMGR: Block manager shutting down
2020-01-22 16:05:08.294 [INF] SRVR: seeder 'testnet-seed-1.decred.org' error: Get https://testnet-seed-1.decred.org/api/addrs?ipversion=0&pver=0&services=1: context canceled
==================
WARNING: DATA RACE
Write at 0x00c00ec2a6a8 by goroutine 33:
  sync.(*WaitGroup).Wait()
      /home/user/apps/go/src/internal/race/race.go:41 +0xee
  main.(*rpcServer).Stop()
      /home/user/projetos/decred/dcrd/rpcwebsocket.go:1294 +0x1b7
  main.(*server).Run()
      /home/user/projetos/decred/dcrd/server.go:2553 +0x3cd
  main.dcrdMain.func5()
      /home/user/projetos/decred/dcrd/dcrd.go:194 +0x4c

Previous read at 0x00c00ec2a6a8 by goroutine 36:
  sync.(*WaitGroup).Add()
      /home/user/apps/go/src/internal/race/race.go:37 +0x18d
  main.(*wsNotificationManager).Start()
      /home/user/projetos/decred/dcrd/rpcwebsocket.go:1286 +0x4b
  main.(*rpcServer).Start()
      /home/user/projetos/decred/dcrd/rpcserver.go:5499 +0x42a
  main.(*server).Run.func1()
      /home/user/projetos/decred/dcrd/server.go:2514 +0x59
```

I believe this is due to the `wsNotificationManager.Start()` being called from a different goroutine than `wsNotificationManager.Stop()` (starting [here](https://github.com/decred/dcrd/blob/e3547d0822db291479ea67a914feaa70806f0828/server.go#L2513-L2514)).

A way to solve this would be to have all subsystems start their own handling goroutines (instead of having `server.Run()` do it) while the server itself calls each `Start()` serially. 

This is a low priority race since it involves node shutdown and can't really cause data loss. I'm creating the issue just to have it documented.

Minimal repro:

```
$ cat /tmp/wg.go
package main

import "sync"

func main() {
	var wg sync.WaitGroup
	for {
		go func() {
			wg.Add(1)
			wg.Done()
		}()
		wg.Wait()
	}
}
$ go run -race /tmp/wg.go
==================
WARNING: DATA RACE
Write at 0x00c000018078 by main goroutine:
  sync.(*WaitGroup).Wait()
      /usr/local/go/src/internal/race/race.go:41 +0xee
  main.main()
      /tmp/wg.go:12 +0x99

Previous read at 0x00c000018078 by goroutine 28:
  sync.(*WaitGroup).Add()
      /usr/local/go/src/internal/race/race.go:37 +0x18d
  main.main.func1()
      /tmp/wg.go:9 +0x41

Goroutine 28 (finished) created at:
  main.main()
      /tmp/wg.go:8 +0x8b
==================
==================
WARNING: DATA RACE
Read at 0x00c000018078 by goroutine 29:
  sync.(*WaitGroup).Add()
      /usr/local/go/src/internal/race/race.go:37 +0x18d
  main.main.func1()
      /tmp/wg.go:9 +0x41

Previous write at 0x00c000018078 by main goroutine:
  sync.(*WaitGroup).Wait()
      /usr/local/go/src/internal/race/race.go:41 +0xee
  main.main()
      /tmp/wg.go:12 +0x99

Goroutine 29 (running) created at:
  main.main()
      /tmp/wg.go:8 +0x8b
==================
panic: sync: WaitGroup is reused before previous Wait has returned

goroutine 1 [running]:
sync.(*WaitGroup).Wait(0xc000018070)
	/usr/local/go/src/sync/waitgroup.go:132 +0x142
main.main()
	/tmp/wg.go:12 +0x9a
exit status 2
```

--- a/rpcserver.go
+++ b/rpcserver.go
@@ -4853,8 +4853,6 @@ func handleVersion(_ context.Context, s *rpcServer, cmd interface{}) (interface{
 
 // rpcServer provides a concurrent safe RPC server to a chain server.
 type rpcServer struct {
-	started                int32
-	shutdown               int32
 	cfg                    rpcserverConfig
 	authsha                [sha256.Size]byte
 	limitauthsha           [sha256.Size]byte
@@ -4923,13 +4921,8 @@ func (s *rpcServer) writeHTTPResponseHeaders(req *http.Request, headers http.Hea
 	return err
 }
 
-// Stop is used by server.go to stop the rpc listener.
-func (s *rpcServer) Stop() error {
-	if atomic.AddInt32(&s.shutdown, 1) != 1 {
-		rpcsLog.Infof("RPC server is already in the process of " +
-			"shutting down")
-		return nil
-	}
+// shutdown terminates the processes of the rpc server.
+func (s *rpcServer) shutdown() error {
 	rpcsLog.Warnf("RPC server shutting down")
 	for _, listener := range s.cfg.Listeners {
 		err := listener.Close()
@@ -4938,8 +4931,6 @@ func (s *rpcServer) Stop() error {
 			return err
 		}
 	}
-	s.ntfnMgr.Shutdown()
-	s.ntfnMgr.WaitForShutdown()
 	s.wg.Wait()
 	rpcsLog.Infof("RPC server shutdown complete")
 	return nil
@@ -5173,9 +5164,11 @@ func (s *rpcServer) processRequest(ctx context.Context, request *dcrjson.Request
 }
 
 // jsonRPCRead handles reading and responding to RPC messages.
-func (s *rpcServer) jsonRPCRead(w http.ResponseWriter, r *http.Request, isAdmin bool) {
-	if atomic.LoadInt32(&s.shutdown) != 0 {
+func (s *rpcServer) jsonRPCRead(sCtx context.Context, w http.ResponseWriter, r *http.Request, isAdmin bool) {
+	select {
+	case <-sCtx.Done():
 		return
+	default:
 	}
 
 	// Read and close the JSON-RPC request body from the caller.
@@ -5219,7 +5212,7 @@ func (s *rpcServer) jsonRPCRead(w http.ResponseWriter, r *http.Request, isAdmin 
 	conn.SetReadDeadline(timeZeroVal)
 	// Setup a close notifier.  Since the connection is hijacked,
 	// the CloseNotifier on the ResponseWriter is not available.
-	ctx, cancel := context.WithCancel(context.Background())
+	ctx, cancel := context.WithCancel(sCtx)
 	defer cancel()
 
 	go func() {
@@ -5402,13 +5395,8 @@ func jsonAuthFail(w http.ResponseWriter) {
 	http.Error(w, "401 Unauthorized.", http.StatusUnauthorized)
 }
 
-// Start is used by server.go to start the rpc listener.
-func (s *rpcServer) Start() {
-	if atomic.AddInt32(&s.started, 1) != 1 {
-		return
-	}
-
-	rpcsLog.Trace("Starting RPC server")
+// route sets up the endpoints of the rpc server.
+func (s *rpcServer) route(ctx context.Context) *http.Server {
 	rpcServeMux := http.NewServeMux()
 	httpServer := &http.Server{
 		Handler: rpcServeMux,
@@ -5437,7 +5425,7 @@ func (s *rpcServer) Start() {
 		}
 
 		// Read and respond to the request.
-		s.jsonRPCRead(w, r, isAdmin)
+		s.jsonRPCRead(ctx, w, r, isAdmin)
 	})
 
 	// Websocket endpoint.
@@ -5477,20 +5465,30 @@ func (s *rpcServer) Start() {
 		})
 		s.WebsocketHandler(ws, r.RemoteAddr, authenticated, isAdmin)
 	})
+	return httpServer
+}
 
+// Run starts the rpc server and its listeners. It blocks until the
+// provided context is cancelled.
+func (s *rpcServer) Run(ctx context.Context) {
+	rpcsLog.Trace("Starting RPC server")
+	server := s.route(ctx)
 	for _, listener := range s.cfg.Listeners {
 		s.wg.Add(1)
 		go func(listener net.Listener) {
-			rpcsLog.Infof("RPC server listening on %s",
-				listener.Addr())
-			httpServer.Serve(listener)
-			rpcsLog.Tracef("RPC listener done for %s",
-				listener.Addr())
+			rpcsLog.Infof("RPC server listening on %s", listener.Addr())
+			server.Serve(listener)
+			rpcsLog.Tracef("RPC listener done for %s", listener.Addr())
 			s.wg.Done()
 		}(listener)
 	}
 
-	s.ntfnMgr.Start()
+	s.ntfnMgr.Run(ctx)
+	err := s.shutdown()
+	if err != nil {
+		rpcsLog.Error(err)
+		return
+	}
 }
 
 // genCertPair generates a key/cert pair to the paths provided.

--- a/rpcwebsocket.go
+++ b/rpcwebsocket.go
@@ -144,23 +144,26 @@ type wsNotificationManager struct {
 	numClients chan int
 
 	// Shutdown handling
-	wg   sync.WaitGroup
-	quit chan struct{}
+	wg  sync.WaitGroup
+	ctx context.Context
+	mtx sync.RWMutex
 }
 
-// queueHandler manages a queue of empty interfaces, reading from in and
-// sending the oldest unsent to out.  This handler stops when either of the
-// in or quit channels are closed, and closes out before returning, without
-// waiting to send any variables still remaining in the queue.
-func queueHandler(in <-chan interface{}, out chan<- interface{}, quit <-chan struct{}) {
+// queueHandler maintains a queue of notifications and notification handler
+// control messages. The handler stops when the input channel is closed or a
+// context cancellation signal is received.
+func (m *wsNotificationManager) queueHandler() {
 	var q []interface{}
 	var dequeue chan<- interface{}
-	skipQueue := out
+	skipQueue := m.notificationMsgs
 	var next interface{}
 out:
 	for {
 		select {
-		case n, ok := <-in:
+		case <-m.ctx.Done():
+			break out
+
+		case n, ok := <-m.queueNotification:
 			if !ok {
 				// Sender closed input channel.
 				break out
@@ -173,7 +176,7 @@ out:
 			case skipQueue <- n:
 			default:
 				q = append(q, n)
-				dequeue = out
+				dequeue = m.notificationMsgs
 				skipQueue = nil
 				next = q[0]
 			}
@@ -184,22 +187,14 @@ out:
 			q = q[:len(q)-1]
 			if len(q) == 0 {
 				dequeue = nil
-				skipQueue = out
+				skipQueue = m.notificationMsgs
 			} else {
 				next = q[0]
 			}
-
-		case <-quit:
-			break out
 		}
 	}
-	close(out)
-}
 
-// queueHandler maintains a queue of notifications and notification handler
-// control messages.
-func (m *wsNotificationManager) queueHandler() {
-	queueHandler(m.queueNotification, m.notificationMsgs, m.quit)
+	close(m.notificationMsgs)
 	m.wg.Done()
 }
 
@@ -207,66 +202,110 @@ func (m *wsNotificationManager) queueHandler() {
 // to the notification manager for block and transaction notification
 // processing.
 func (m *wsNotificationManager) NotifyBlockConnected(block *dcrutil.Block) {
+	m.mtx.RLock()
+	if m.ctx == nil {
+		// Notification manager not started yet.
+		m.mtx.RUnlock()
+		return
+	}
+	ctx := m.ctx
+	m.mtx.RUnlock()
+
 	// As NotifyBlockConnected will be called by the block manager
 	// and the RPC server may no longer be running, use a select
 	// statement to unblock enqueuing the notification once the RPC
 	// server has begun shutting down.
 	select {
 	case m.queueNotification <- (*notificationBlockConnected)(block):
-	case <-m.quit:
+	case <-ctx.Done():
 	}
 }
 
 // NotifyBlockDisconnected passes a block disconnected from the best chain
 // to the notification manager for block notification processing.
 func (m *wsNotificationManager) NotifyBlockDisconnected(block *dcrutil.Block) {
+	m.mtx.RLock()
+	if m.ctx == nil {
+		// Notification manager not started yet.
+		m.mtx.RUnlock()
+		return
+	}
+	ctx := m.ctx
+	m.mtx.RUnlock()
+
 	// As NotifyBlockDisconnected will be called by the block manager
 	// and the RPC server may no longer be running, use a select
 	// statement to unblock enqueuing the notification once the RPC
 	// server has begun shutting down.
 	select {
 	case m.queueNotification <- (*notificationBlockDisconnected)(block):
-	case <-m.quit:
+	case <-ctx.Done():
 	}
 }
 
 // NotifyWork passes new mining work to the notification manager
 // for block notification processing.
 func (m *wsNotificationManager) NotifyWork(templateNtfn *TemplateNtfn) {
+	m.mtx.RLock()
+	if m.ctx == nil {
+		// Notification manager not started yet.
+		m.mtx.RUnlock()
+		return
+	}
+	ctx := m.ctx
+	m.mtx.RUnlock()
+
 	// As NotifyWork will be called by the block manager and the
 	// RPC server may no longer be running, use a select statement
 	// to unblock enqueuing the notification once the RPC server
 	// has begun shutting down.
 	select {
 	case m.queueNotification <- (*notificationWork)(templateNtfn):
-	case <-m.quit:
+	case <-ctx.Done():
 	}
 }
 
 // NotifyReorganization passes a blockchain reorganization notification for
 // reorganization notification processing.
 func (m *wsNotificationManager) NotifyReorganization(rd *blockchain.ReorganizationNtfnsData) {
+	m.mtx.RLock()
+	if m.ctx == nil {
+		// Notification manager not started yet.
+		m.mtx.RUnlock()
+		return
+	}
+	ctx := m.ctx
+	m.mtx.RUnlock()
+
 	// As NotifyReorganization will be called by the block manager
 	// and the RPC server may no longer be running, use a select
 	// statement to unblock enqueuing the notification once the RPC
 	// server has begun shutting down.
 	select {
 	case m.queueNotification <- (*notificationReorganization)(rd):
-	case <-m.quit:
+	case <-ctx.Done():
 	}
 }
 
 // NotifyWinningTickets passes newly winning tickets for an incoming block
 // to the notification manager for further processing.
-func (m *wsNotificationManager) NotifyWinningTickets(
-	wtnd *WinningTicketsNtfnData) {
+func (m *wsNotificationManager) NotifyWinningTickets(wtnd *WinningTicketsNtfnData) {
+	m.mtx.RLock()
+	if m.ctx == nil {
+		// Notification manager not started yet.
+		m.mtx.RUnlock()
+		return
+	}
+	ctx := m.ctx
+	m.mtx.RUnlock()
+
 	// As NotifyWinningTickets will be called by the block manager
 	// and the RPC server may no longer be running, use a select
 	// statement to unblock enqueuing the notification once the RPC
 	// server has begun shutting down.
 	select {
 	case m.queueNotification <- (*notificationWinningTickets)(wtnd):
-	case <-m.quit:
+	case <-ctx.Done():
 	}
 }
 
@@ -275,13 +314,22 @@ func (m *wsNotificationManager) NotifyWinningTickets(
 // notification processing.
 func (m *wsNotificationManager) NotifySpentAndMissedTickets(
 	tnd *blockchain.TicketNotificationsData) {
+	m.mtx.RLock()
+	if m.ctx == nil {
+		// Notification manager not started yet.
+		m.mtx.RUnlock()
+		return
+	}
+	ctx := m.ctx
+	m.mtx.RUnlock()
+
 	// As NotifySpentAndMissedTickets will be called by the block manager
 	// and the RPC server may no longer be running, use a select
 	// statement to unblock enqueuing the notification once the RPC
 	// server has begun shutting down.
 	select {
 	case m.queueNotification <- (*notificationSpentAndMissedTickets)(tnd):
-	case <-m.quit:
+	case <-ctx.Done():
 	}
 }
 
@@ -289,13 +337,22 @@ func (m *wsNotificationManager) NotifySpentAndMissedTickets(
 // chain to the notification manager for block notification processing.
 func (m *wsNotificationManager) NotifyNewTickets(
 	tnd *blockchain.TicketNotificationsData) {
+	m.mtx.RLock()
+	if m.ctx == nil {
+		// Notification manager not started yet.
+		m.mtx.RUnlock()
+		return
+	}
+	ctx := m.ctx
+	m.mtx.RUnlock()
+
 	// As NotifyNewTickets will be called by the block manager
 	// and the RPC server may no longer be running, use a select
 	// statement to unblock enqueuing the notification once the RPC
 	// server has begun shutting down.
 	select {
 	case m.queueNotification <- (*notificationNewTickets)(tnd):
-	case <-m.quit:
+	case <-ctx.Done():
 	}
 }
 
@@ -303,13 +360,22 @@ func (m *wsNotificationManager) NotifyNewTickets(
 // chain to the notification manager for block notification processing.
 func (m *wsNotificationManager) NotifyStakeDifficulty(
 	stnd *StakeDifficultyNtfnData) {
+	m.mtx.RLock()
+	if m.ctx == nil {
+		// Notification manager not started yet.
+		m.mtx.RUnlock()
+		return
+	}
+	ctx := m.ctx
+	m.mtx.RUnlock()
+
 	// As NotifyNewTickets will be called by the block manager
 	// and the RPC server may no longer be running, use a select
 	// statement to unblock enqueuing the notification once the RPC
 	// server has begun shutting down.
 	select {
 	case m.queueNotification <- (*notificationStakeDifficulty)(stnd):
-	case <-m.quit:
+	case <-ctx.Done():
 	}
 }
 
@@ -318,6 +384,15 @@ func (m *wsNotificationManager) NotifyStakeDifficulty(
 // isNew is true, the tx is a new transaction, rather than one
 // added to the mempool during a reorg.
 func (m *wsNotificationManager) NotifyMempoolTx(tx *dcrutil.Tx, isNew bool) {
+	m.mtx.RLock()
+	if m.ctx == nil {
+		// Notification manager not started yet.
+		m.mtx.RUnlock()
+		return
+	}
+	ctx := m.ctx
+	m.mtx.RUnlock()
+
 	n := &notificationTxAcceptedByMempool{
 		isNew: isNew,
 		tx:    tx,
@@ -329,7 +404,7 @@ func (m *wsNotificationManager) NotifyMempoolTx(tx *dcrutil.Tx, isNew bool) {
 	// shutting down.
 	select {
 	case m.queueNotification <- n:
-	case <-m.quit:
+	case <-ctx.Done():
 	}
 }
 
@@ -529,6 +604,10 @@ func (m *wsNotificationManager) notificationHandler() {
 out:
 	for {
 		select {
+		case <-m.ctx.Done():
+			// RPC server shutdown.
+			break out
+
 		case n, ok := <-m.notificationMsgs:
 			if !ok {
 				// queueHandler quit.
@@ -657,10 +736,6 @@ out:
 			}
 
 		case m.numClients <- len(clients):
-
-		case <-m.quit:
-			// RPC server shutting down.
-			break out
 		}
 	}
 
@@ -672,9 +747,18 @@ out:
 
 // NumClients returns the number of clients actively being served.
 func (m *wsNotificationManager) NumClients() (n int) {
+	m.mtx.RLock()
+	if m.ctx == nil {
+		// Notification manager not started yet.
+		m.mtx.RUnlock()
+		return
+	}
+	ctx := m.ctx
+	m.mtx.RUnlock()
+
 	select {
 	case n = <-m.numClients:
-	case <-m.quit: // Use default n (0) if server has shut down.
+	case <-ctx.Done(): // Use default n (0) if server has shut down.
 	}
 	return
 }
@@ -1274,30 +1358,33 @@ func (m *wsNotificationManager) AddClient(wsc *wsClient) {
 // RemoveClient removes the passed websocket client and all notifications
 // registered for it.
 func (m *wsNotificationManager) RemoveClient(wsc *wsClient) {
+	m.mtx.RLock()
+	if m.ctx == nil {
+		// Notification manager not started yet.
+		m.mtx.RUnlock()
+		return
+	}
+	ctx := m.ctx
+	m.mtx.RUnlock()
+
 	select {
 	case m.queueNotification <- (*notificationUnregisterClient)(wsc):
-	case <-m.quit:
+	case <-ctx.Done():
 	}
 }
 
-// Start starts the goroutines required for the manager to queue and process
-// websocket client notifications.
-func (m *wsNotificationManager) Start() {
+// Run starts the goroutines required for the manager to queue and process
+// websocket client notifications. It blocks until the provided context is
+// cancelled.
+func (m *wsNotificationManager) Run(ctx context.Context) {
+	m.mtx.Lock()
+	m.ctx = ctx
+	m.mtx.Unlock()
+
 	m.wg.Add(2)
 	go m.queueHandler()
 	go m.notificationHandler()
-}
-
-// WaitForShutdown blocks until all notification manager goroutines have
-// finished.
-func (m *wsNotificationManager) WaitForShutdown() {
 	m.wg.Wait()
-}
-
-// Shutdown shuts down the manager, stopping the notification queue and
-// notification handler goroutines.
-func (m *wsNotificationManager) Shutdown() {
-	close(m.quit)
 }
 
 // newWsNotificationManager returns a new notification manager ready for use.
@@ -1308,7 +1395,6 @@ func newWsNotificationManager(server *rpcServer) *wsNotificationManager {
 		queueNotification: make(chan interface{}),
 		notificationMsgs:  make(chan interface{}),
 		numClients:        make(chan int),
-		quit:              make(chan struct{}),
 	}
 }
 

--- a/server.go
+++ b/server.go
@@ -2542,8 +2542,7 @@ func (s *server) Run(ctx context.Context) {
 
 		s.wg.Add(1)
 		go func(s *server) {
-			s.rpcServer.Start()
-			s.rpcServer.wg.Wait()
+			s.rpcServer.Run(serverCtx)
 			s.wg.Done()
 		}(s)
 	}
@@ -2577,11 +2576,6 @@ func (s *server) Run(ctx context.Context) {
 	// Stop the CPU miner if needed.
 	if cfg.Generate && s.cpuMiner != nil {
 		s.cpuMiner.Stop()
-	}
-
-	// Shutdown the RPC server if it's not disabled.
-	if !cfg.DisableRPC && s.rpcServer != nil {
-		s.rpcServer.Stop()
 	}
 
 	s.feeEstimator.Close()

